### ISSUE TITLE
Fix Backup purchase flow for site-only connection

### ISSUE THREAD

**Issue by leogermani** [Automattic/jetpack#23171]
Depends on #23170

When the Backup plugin triggers the connection flow from the Backup plugin, there's a special rule that will redirect the user to the checkout page. (see https://github.com/Automattic/wp-calypso/pull/60468)

We need to 

- [ ] Remove this special behavior from Calypso
- [ ] Make the Flow from the plugin use the `unlinked=1` approach, to be consistent with all other flows
- [ ] Also apply the same flow to the My Jetpack activation flow.

**Comment by kangzj** [comment]
Hi @leogermani Search has something similar as Backup and we want to align the purchase flow, because luckily the Search plugin is not in the wild yet (first release at the end of May).

However I have a question, I'm using `ConnectScreenRequiredPlan` as the connection screen, which doesn't support passed in handler for the CTA button. And that's why I implemented the same special behaviour in Calypso to redirect to checkout after connection, otherwise the Jetpack product list really doesn't make much sense to a Search only user.

https://github.com/Automattic/jetpack/blob/daf16556cc805df342914e860aa4d1d2001b6161/projects/packages/search/src/dashboard/components/dashboard/use-connection.jsx#L36

PS:
And I'm currently working on a upsell page for Search and implemented the recommended behaviour:

https://github.com/Automattic/jetpack/pull/23916/files#diff-9b1e7694de1d864eda123df47e5a96a0f96263397047009ef1644078ab916271R35

Thanks in advance 👍

**Comment by leogermani** [comment]
Hi @kangzj thanks for reaching out.

This is not completely sorted out.

I think the way to go here would be to change the behavior of `ConnectScreenRequiredPlan` component.

Instead of simply connecting, the component should:

* Connect at a site level via ajax
* Redirect the user to the checkout page of the required plan passing the `unlinked=1` as a parameter

This would in fact make the user land in the authorization screen and only then redirect them to the checkout page.

I just think it's not a good pattern having to add special handlers in calypso for each product. If we did that it would work for every product. And if we need to change the product for a specific plugin, it's also easy. (If we use a redirect source, we can even change it from our Redirect Manager UI, no need to code anything)

My only concern is how to handle this with the current usage that Backup makes of this component. Do you have any thought? I'll try to look a bit more into that tomorrow

**Comment by kangzj** [comment]
Thanks for your thoughts on this @leogermani 

> I just think it's not a good pattern having to add special handlers in calypso for each product.

Indeed totally agree.

> Instead of simply connecting, the component should:
>
>    ...
>    Redirect the user to the checkout page of the required plan passing the unlinked=1 as a parameter

I'm not sure which plugins are using the components other than Search and backup. Just thinking, could we make it the default behaviour for the CTA button - at the end of the day, the button is for plan purchasing eh? If users just want to connect, then they could click the `Log in to get started` link (which is a bit hard to notice tho) to connect their site.

The only problem is, if a user already has search subscription, the checkout link would try to add another one (?) while the existing auth logic would redirect user back to the site. Tried to look into it, manually composed a checkout link with `unlinked=1` but no luck:

![image](https://user-images.githubusercontent.com/1425433/163287728-ca5ced74-abad-4b4a-9e87-fab3c68fe76b.png)

**Comment by leogermani** [comment]
> I'm not sure which plugins are using the components other than Search and backup. Just thinking, could we make it the default behaviour for the CTA button 

Yes we should be able to adapt it in a way that Backup keeps its current behavior

> The only problem is, if a user already has search subscription, the checkout link would try to add another one (?) 

What we do in My Jetpack is that we check if the site has the plan beforehand. If it has, we just connect, no need to go to the checkout flow.

We even don't show the prices or anything, because we know the site already has the plan and we don't want to confuse the user by making him think that they will have to pay for it again if they click the button.

This works well because the plans are tied to the site, and not to the user, so we are able to fetch this info even before the connection is established

**Comment by leogermani** [comment]
@kangzj @retrofox and I discussed this a bit and we will think on the best way to implement it in a way it will be easy for consumers to implement this flow.

I've opened a dedicated issue for this: https://github.com/Automattic/jetpack/issues/23972

Let's aim to have this ready in the next couple of weeks so we can integrate Search with it and get rid of the Calypso exception

**Comment by leogermani** [comment]
@kangzj We have this Hook now to help achieving the desired flow:

https://github.com/Automattic/jetpack/tree/master/projects/js-packages/connection/hooks/use-product-checkout-workflow

* Connects the site (if not connected)
* Redirects the user to the checkout page with `unlinked=1` if user not linked

It can be used anywhere in the UI, not only in the initial connection flow. We are using it in 2 places in the Protect plugin.

Do you think this helps you achieve this flow in Search?

We could integrate it with the `ConnectScreenRequiredPlan` now. I'll work on this

**Comment by kangzj** [comment]
> Do you think this helps you achieve this flow in Search?

hi @leogermani hooray, this is exactly what we want! thank you and @retrofox for working on that.

> We could integrate it with the ConnectScreenRequiredPlan now. I'll work on this

That'll be fantastic! When this is implemented, I'll reverse the changes for the special redirection behavior from https://github.com/Automattic/wp-calypso/pull/62199.

Thanks again for all your efforts 👍

**Comment by kangzj** [comment]
For late comers, Jetpack Search now leverages the unlinked checkout flow.

**Comment by bizanimesh** [comment]
We just want to check - Will this one fix Automattic/jpop-issues#7616 issue as well?

**Comment by bobmatyas** [comment]
@leogermani Will this one fix https://github.com/Automattic/jpop-issues/issues/7616?

**Comment by leogermani** [comment]
I don't think this will fix that issue by itself.

I will depend on how WPCOM responds when My Jetpack asks "Does this site already have search?"

It does so by reaching `/sites/$blog_id/jetpack-search/plan` endpoint. If this endpoint takes the legacy plans into consideration, then the user won't be asked to buy a new one. -> https://github.com/Automattic/jetpack/blob/trunk/projects/packages/my-jetpack/src/products/class-search.php#L182

Same thing goes for Backup, that reaches `/sites/%d/rewind`. -> https://github.com/Automattic/jetpack/blob/trunk/projects/packages/my-jetpack/src/products/class-backup.php#L152

cc @oskosk @lhkowalski

**Comment by jwebbdev** [comment]
I'm working on this in https://github.com/Automattic/jetpack/pull/26943 - And things seem to work similar to how it's described by search.

Additionally we have a case to handle where the user has purchased a siteless key.  I wonder if this might be best handled similar to how @kangzj mentioned:
> If users just want to connect, then they could click the Log in to get started link (which is a bit hard to notice tho) to connect their site.

The initial connection would then allow My Jetpack to load, which gives access to the License Activation.
